agent/otel.go

```diff
@@ -279,6 +288,166 @@ func hostResource(ctx context.Context) (*resource.Resource, error) {
+
+	var resourceSpans []*tracepb.ResourceSpans
+	scanner := bufio.NewScanner(file)
+	scanner.Buffer([]byte{}, maxLineSize)
+	for scanner.Scan() {
+		var traces tracepb.TracesData
+		catcher.Wrap(protojson.Unmarshal(scanner.Bytes(), &traces), "unmarshalling trace")
+		resourceSpans = append(resourceSpans, traces.ResourceSpans...)
+	}
+	if err := scanner.Err(); err != nil {
+		catcher.Wrapf(err, "scanning file '%s'", fileName)
+	}
+
+	if err = fixBinaryIDs(resourceSpans); err != nil {
+		return nil, errors.Wrapf(err, "fixing binary IDs for '%s'", fileName)
+	}
+
+	return resourceSpans, catcher.Resolve()
+}
+
+// fixBinaryIDs fixes every trace and span id in resourceSpans. These IDs are encoded
+// as hex strings in the source file because that's how the [OTel JSON protobuf encoding] is defined
+// but [protojson] assumes they're encoded with base64 encoding since that's the [standard JSON encoding].
+// We need to iterate through the spans and fix them.
+//
+// [OTel JSON protobuf encoding]: https://opentelemetry.io/docs/specs/otel/protocol/otlp/#json-protobuf-encoding
+// [standard JSON encoding]: https://protobuf.dev/programming-guides/proto3/#json
+func fixBinaryIDs(resourceSpans []*tracepb.ResourceSpans) error {
+	catcher := grip.NewBasicCatcher()
+	for _, rs := range resourceSpans {
+		for _, ss := range rs.ScopeSpans {
+			for _, span := range ss.Spans {
+				catcher.Wrap(fixSpan(span), "fixing span")
+				for _, spanLink := range span.Links {
+					catcher.Add(fixSpanLink(spanLink))
+				}
+			}
+		}
+	}
+
+	return catcher.Resolve()
+}
+
+func fixSpan(span *tracepb.Span) error {
+	traceIDHex, err := fixBinaryID(span.TraceId)
+	if err != nil {
+		return errors.Wrap(err, "fixing trace id")
+	}
+	spanIDHex, err := fixBinaryID(span.SpanId)
+	if err != nil {
+		return errors.Wrap(err, "fixing span id")
+	}
+	parentSpanIDHex, err := fixBinaryID(span.ParentSpanId)
+	if err != nil {
+		return errors.Wrap(err, "fixing parent span id")
+	}
+
+	span.TraceId = traceIDHex
+	span.SpanId = spanIDHex
+	span.ParentSpanId = parentSpanIDHex
+	return nil
+}
+
+func fixSpanLink(spanLink *tracepb.Span_Link) error {
+	traceIDHex, err := fixBinaryID(spanLink.TraceId)
+	if err != nil {
+		return errors.Wrap(err, "fixing trace id")
+	}
+	spanIDHex, err := fixBinaryID(spanLink.SpanId)
+	if err != nil {
+		return errors.Wrap(err, "fixing span id")
+	}
+
+	spanLink.TraceId = traceIDHex
+	spanLink.SpanId = spanIDHex
+	return nil
+}
+
+// fixBinaryID recovers the original hex string id and decodes it back
+// into []byte. The unmarshaller decoded the string as a base64 encoded
+// string so we encode it back to the string and decode it again to []byte.
+func fixBinaryID(id []byte) ([]byte, error) {
+	idHex := base64.StdEncoding.EncodeToString(id)
+	return hex.DecodeString(idHex)
+}
+
+// getTraceFiles returns the full path of all the files in the [traceSuffix] directory
+// under the task's working directory.
+func getTraceFiles(taskDir string) ([]string, error) {
```

Comment:
@julianedwards do we want to model this similarly to `build/TestLogs` - should this be `build/OTelTraces` or something?

Reply:
I think that is a good idea!